### PR TITLE
MB-2329 Revoke sessions via admin interface

### PR DESCRIPTION
**Pull request by monfresh**
One of the STIG requirements is the ability to revoke individual
sessions, and because we don't have access to the Redis server in
prod, we need to provide a way for someone to do this via the admin
website.

**Setup**:
`make db_dev_migrate`
`make redis_pull`

**To test session revocation:**

1. Sign in at http://milmovelocal:3000/ in one browser tab
2. In another browser tab, sign in at http://adminlocal:3000/ (You
might need to create yourself as an admin first via
http://adminlocal:3000/devlocal-auth/login)
3. Once in the admin site, click on Search User by ID on the left
sidebar
4. Enter your UUID and click the Search button. You can find your UUID
in the server logs in the middleware/request_logger.go:85 line under
the key `user-id`.
5. Once your details appear in the user show page, click on Edit in
the top right
6. Select "Yes" from the dropdown for "Revoke mil session"
7. Go back to the tab where you were signed into the mil app and
refresh it.
8.Verify that you are logged out, and that the session token has been
removed from the `current_mil_session_id` field in the Users table,
and from Redis (accessed via `redis-dev`, then list all keys with
`KEYS *`)

## Description

Explain a little about the changes at a high level.

## Reviewer Notes

Is there anything you would like reviewers to give additional scrutiny?

## Setup

Add any steps or code to run in this section to help others prepare to run your code:

```sh
echo "Code goes here"
```

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/wiki/backend#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/wiki/backend#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/wiki/migrate-the-database#zero-downtime-migrations))
  * [ ] Have been communicated to #dp3-engineering
  * [ ] Secure migrations have been tested following the instructions in our [docs](https://github.com/transcom/mymove/wiki/migrate-the-database#secure-migrations)
* [ ] There are no aXe warnings for UI.
* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](tbd) for this change
* [this article](tbd) explains more about the approach used.

## Screenshots

If this PR makes visible UI changes, an image of the finished UI can help reviewers and casual
observers understand the context of the changes. A before image is optional and
can be included at the submitter's discretion.

Consider using an animated image to show an entire workflow instead of using multiple images. You may want to use GIPHY CAPTURE for this! 📸

_Please frame screenshots to show enough useful context but also highlight the affected regions._

### CODE REVIEW

File: pkg/handlers/adminapi/users.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	userID, err := uuid.FromString(params.UserID.String())
+	if err != nil {
+		logger.Error(fmt.Sprintf("UUID Parsing for %s", params.UserID.String()), zap.Error(err))
+	}
+
+	sessionStore := h.SessionManager(session).Store
+	updatedUser, validationErrors, revokeErr := h.UserSessionRevocation.RevokeUserSession(userID, payload, sessionStore)
+	if revokeErr != nil || validationErrors != nil {
+		fmt.Printf("%#v", validationErrors)
+		logger.Error("Error saving user", zap.Error(revokeErr))
```
Comment (duncan-truss) [inline pkg/handlers/adminapi/users.go:69]: May not want that debug statement in there and the error message doesn't seem to describe that it's a delete operation right?

File: pkg/services/user/user_session_revocation.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	filters := []services.QueryFilter{query.NewQueryFilter("id", "=", id.String())}
+	err := o.builder.FetchOne(&foundUser, filters)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var currentSessionID string
+
+	if payload.RevokeAdminSession != nil && *payload.RevokeAdminSession == true {
```
Comment (duncan-truss) [inline pkg/services/user/user_session_revocation.go:32]: Is there a case where the payload will be set to true for more than one user type at a time?

File: pkg/services/user/user_session_revocation.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	if payload.RevokeOfficeSession != nil && *payload.RevokeOfficeSession == true {
+		currentSessionID = foundUser.CurrentOfficeSessionID
+		deleteSessionIDFromRedis(currentSessionID, sessionStore)
+	}
+
+	if payload.RevokeMilSession != nil && *payload.RevokeMilSession == true {
+		currentSessionID = foundUser.CurrentMilSessionID
+		deleteSessionIDFromRedis(currentSessionID, sessionStore)
```
Comment (duncan-truss) [inline pkg/services/user/user_session_revocation.go:44]: I think we should check and log for an error coming back from `deleteSessionIDFromRedis`

File: pkg/handlers/adminapi/users.go (hunk 1 above)
Comment (monfresh) [reply pkg/handlers/adminapi/users.go:69]: I added this logger statement to follow what we're doing in other parts of the admin interface that make updates to users. I agree that the error message could be improved. I will update it.

File: pkg/services/user/user_session_revocation.go (hunk 2 above)
Comment (monfresh) [reply pkg/services/user/user_session_revocation.go:32]: Yes. There is an integration test for that in adminapi/users_test.go on line 124, but that only tests the DB changes. I will add a new Redis unit test in user_session_revocation_test.go to test the scenario where there are multiple sessions that need to be revoked.

File: pkg/services/user/user_session_revocation.go (hunk 3 above)
Comment (monfresh) [reply pkg/services/user/user_session_revocation.go:44]: Yes! Good catch!

File: pkg/services/user/user_session_revocation.go (hunk 2 above)
Comment (duncan-truss) [reply pkg/services/user/user_session_revocation.go:32]: Yep it seemed possible from the interface but the redis tests were all single so it threw me off

**Pull request merged**